src/GeniusYield/TxBuilder/Common.hs

```diff
@@ -100,27 +101,40 @@ buildTxCore ss eh pp ps cstrat ownUtxoUpdateF addrs change collateral action = d
 
             refInsUtxos <- utxosAtTxOutRefs $ gyTxSkeletonRefInsToList gytxRefIns
 
-            -- Build the transaction.
-            buildUnsignedTxBody
-                (buildEnvWith ownUtxos' (gyTxSkeletonRefInsSet gytxRefIns))
-                cstrat
-                gyTxInsDetailed
-                gytxOuts
-                refInsUtxos
-                gytxMint'
-                gytxInvalidBefore
-                gytxInvalidAfter
-                gytxSigs
+            -- This operation is `O(n)` where `n` denotes the number of UTxOs in `ownUtxos'`.
+            mCollateralUtxo <-
+              maybe
+                ( return $
+                    find
+                      (\u -> utxoValue u `valueGreaterOrEqual` collateralValue)  -- Keeping it simple.
+                      (utxosToList ownUtxos')
+
+                ) (fmap Just . utxoAtTxOutRef') reservedCollateral
```

Comment:
I'm not completely sure about this, but I think we can use the same collateral for every Tx. I mean, it's something that we already did. I wonder if it makes sense here.

Reply:
@alegadea The motivation behind this PR is to not reserve for collateral. Think for a case where wallet has single UTxO with 1000 ada, reserving for it would mean 0 funds available. So collaterals are now _spendable_ and in such a case we would need to have them computed when building for tx skeleton because, say, in case of parallel tx building, collateral selected for previous build might have been consumed and thus be no longer available for subsequent ones.